File: tests/gpu/score_victory_test.py
"""CIV 6'S SCORE and the turn-limit victory, GPU side — the twin of
tests/cpu/victory/score-victory.test.ts.

`score_lines` counts the install's ScoringLineItems (cpu/data/scoring.ts): the
whole game's era score, civics 3, cities 5, completed districts 2 (a completed
wonder's own district included), citizens 1, Great People earned 5, the
founded religion's beliefs 5, techs 2, completed wonders 15. Past the turn
limit `victory_row` names `leader()`: the highest Score among the majors that
hold a city, ties to the higher line in TieBreakerPriority order, then to the
lower row. No scripted game ties at the limit, so this lane pins the order.

Run: PYTHONIOENCODING=utf-8 python tests/gpu/score_victory_test.py
"""
from __future__ import annotations

import sys
from pathlib import Path

import torch

ROOT = Path(__file__).resolve().parents[2]
sys.path.insert(0, str(ROOT / "gpu"))
sys.path.insert(0, str(Path(__file__).resolve().parent))
from core import load_rules, fixture_paths  # noqa: E402
from warmup import opened, plant_city, warm_base  # noqa: E402

B0 = 0


def fresh(rules, path):
    return warm_base(str(path), lambda: opened(rules, path))


def level(sim) -> None:
    """Every major at the same Score: one city of one citizen, nothing else."""
    for r in range(sim.n_majors):
        sim.civ_civics[B0, r] = False
        sim.civ_techs[B0, r] = False
        sim.era_score[B0, r] = 0
        sim.era_score_past[B0, r] = 0
        sim.civ_gp_earned[B0, r] = 0
        for p in ("civ_follower", "civ_founder", "civ_enhancer"):
            getattr(sim, p)[B0, r] = -1
        alive = sim.city_alive[B0, r, : sim.RC]
        sim.city_pop[B0, r, : sim.RC] = torch.where(alive, torch.ones_like(sim.city_pop[B0, r, : sim.RC]),
                                                    sim.city_pop[B0, r, : sim.RC])


def total(sim, r: int) -> int:
    return int(sim.score_lines(r)[B0].sum())


def free_ring(sim, centre: int, n: int) -> list[int]:
    """`n` bare plots one step from `centre` — no district, no wonder, no city."""
    ok = ((sim.pair_dist[centre] == 1) & (sim.district[B0] < 0) & (sim.built_wonder[B0] < 0)
          & (sim.centre_slot_at[B0] < 0))
    ts = ok.nonzero(as_tuple=True)[0].tolist()
    assert len(ts) >= n, f"only {len(ts)} bare plots beside tile {centre}"
    return ts[:n]


def test_counts(rules, path) -> None:
    sim = fresh(rules, path)
    lines = [ln["count"] for ln in sim.rules.scoring]
    assert lines == ["eraScore", "civics", "cities", "districts", "population",
                     "greatPeople", "religion", "techs", "wonders"], lines
    assert [int(ln["value"]) for ln in sim.rules.scoring] == [1, 3, 5, 2, 1, 5, 5, 2, 15]
    r = 0
    plant_city(sim, r)
    cols = sim.city_alive[B0, r, : sim.RC].nonzero(as_tuple=True)[0].tolist()
    assert len(cols) == 2, f"seat {r} holds {len(cols)} cities, wanted 2"
    a, b = cols
    sim.civ_civics[B0, r] = False
    sim.civ_civics[B0, r, :2] = True
    sim.civ_techs[B0, r] = False
    sim.civ_techs[B0, r, :3] = True
    sim.city_pop[B0, r, a] = 3
    sim.city_pop[B0, r, b] = 4
    # a completed Campus counts, a placed Theater Square does not
    campus = next(i for i, d in enumerate(sim.districts_cat) if d["id"] == "CAMPUS")
    theater = next(i for i, d in enumerate(sim.districts_cat) if d["id"] == "THEATER_SQUARE")
    t_campus, t_pyr = free_ring(sim, int(sim.city_center[B0, r, a]), 2)
    t_theater, t_ora = free_ring(sim, int(sim.city_center[B0, r, b]), 2)
    for col, d, t, done in ((a, campus, t_campus, True), (b, theater, t_theater, False)):
        sim.city_dist_tile[B0, r, col, d] = t
        sim.district[B0, t] = d
        sim.district_complete[B0, t] = done
    # a completed wonder counts as a wonder AND as a district; one still under
    # construction counts as neither
    for col, wi, t, done in ((a, 0, t_pyr, True), (b, 1, t_ora, False)):
        sim.city_wonder[B0, r, col, wi] = t
        sim.built_wonder[B0, t] = wi
        sim.built_wonder_complete[B0, t] = done
    sim.civ_gp_earned[B0, r] = 0
    sim.civ_gp_earned[B0, r, 0] = 1
    sim.civ_gp_earned[B0, r, 1] = 1
    # founded and enhanced: three beliefs; the pantheon is not one of them
    sim.civ_pantheon[B0, r] = 0
    sim.civ_follower[B0, r] = 0
    sim.civ_founder[B0, r] = 0
    sim.civ_enhancer[B0, r] = 0
    sim.era_score[B0, r] = 9
    sim.era_score_past[B0, r] = 30
    got = dict(zip(lines, (int(x) for x in sim.score_lines(r)[B0].tolist())))
    want = {"eraScore": 39, "civics": 6, "cities": 10, "districts": 4, "population": 7,
            "greatPeople": 10, "religion": 15, "techs": 6, "wonders": 15}
    assert got == want, f"score lines {got}, wanted {want}"
    print(f"  1 counts OK — {got}")


def test_era_bank(rules, path) -> None:
    sim = fresh(rules, path)
    sim.era_score[B0, 0] = 12
    sim.era_score_past[B0, 0] = 5
    sim.turn = sim._era_len - 1
    sim.step()
    assert int(sim.turn) % sim._era_len == 0
    assert int(sim.era_score[B0, 0]) == 0, "the boundary resets the era's window"
    assert int(sim.era_score_past[B0, 0]) == 17, (
        f"the boundary banks the closed era: {int(sim.era_score_past[B0, 0])}, wanted 17")
    print("  2 era bank OK — 5 + 12 banked at the boundary, the window reset")


def test_ties(rules, path) -> None:
    sim = fresh(rules, path)
    assert sim.n_majors >= 2, "the lane needs two majors"
    level(sim)
    assert total(sim, 0) == total(sim, 1)
    assert int(sim.leader()[B0]) == 0, "a level field goes to the lower row"
    # the era line (1010) outranks the techs (40)
    sim = fresh(rules, path)
    level(sim)
    sim.civ_techs[B0, 0, :2] = True
    sim.era_score[B0, 1] = 4
    assert total(sim, 0) == total(sim, 1)
    assert int(sim.leader()[B0]) == 1, "the era line breaks the tie"
    # the civic line (100) outranks the techs (40)
    sim = fresh(rules, path)
    level(sim)
    sim.civ_techs[B0, 0, :3] = True
    sim.civ_civics[B0, 1, :2] = True
    assert total(sim, 0) == total(sim, 1)
    assert int(sim.leader()[B0]) == 1, "the civic line breaks the tie"
    # a seat with no city cannot win it
    sim = fresh(rules, path)
    level(sim)
    sim.civ_techs[B0, 1, :3] = True
    sim.city_alive[B0, 1] = False
    assert int(sim.leader()[B0]) == 0, "a seat with no city won the Score"
    print("  3 ties OK — era line, then civics over techs, then the lower row; a cityless seat never")


def test_turn_limit(rules, path) -> None:
    sim = fresh(rules, path)
    sim.civ_civics[B0, 1, :4] = True
    sim.turn = int(sim.rules.turn_limit) - 1
    sim.step()
    assert not bool(sim.game_over[B0]) and int(sim.victory_row[B0]) == -1
    sim.step()
    assert bool(sim.game_over[B0]), "the game runs past the turn limit"
    assert int(sim.victory_type[B0]) == 1, f"victory type {int(sim.victory_type[B0])}, wanted 1 (score)"
    assert int(sim.victory_row[B0]) == 1, f"the score victory named row {int(sim.victory_row[B0])}, wanted 1"
    assert int(sim.victory_row[B0]) == int(sim.leader()[B0])
    print("  4 turn limit OK — the game ends on the score and names the leader")


def main() -> int:
    rules = load_rules()
    paths = fixture_paths()
    assert paths, "no fixtures — run `npm run seed && npm run export` first"
    path = paths[0]
    print(f"score_victory_test on {path.name}")
    test_counts(rules, path)
    test_era_bank(rules, path)
    test_ties(rules, path)
    test_turn_limit(rules, path)
    print("score_victory_test OK")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tests/cpu/victory/score-victory.test.ts
import { describe, it, expect } from 'vitest';
import type { GameState } from '../../../cpu/core/types';
import { emptySeat, seatOf } from '../../../cpu/core/seats';
import { endTurn, TURN_LIMIT } from '../../../cpu/core/game';
import { eraBoundary } from '../../../cpu/core/eras';
import { scoreLeader, scoreLines } from '../../../cpu/core/score';
import { SCORING_LINE_ITEMS } from '../../../cpu/data/scoring';
import { ERA_LENGTH } from '../../../cpu/data/seats';
import { GREAT_PEOPLE } from '../../../cpu/data/greatPeople';
import { FOLLOWER_BELIEFS, FOUNDER_BELIEFS, WORSHIP_BELIEFS, ENHANCER_BELIEFS } from '../../../cpu/data/religion';
import { seededGame, makeMap, makeState, settleAt } from '../helpers';

// CIV 6'S SCORE and the turn-limit victory it decides. No scripted game
// reaches the turn limit with a tie, so these pokes pin the counting and the
// tie order. The GPU twin is tests/gpu/score_victory_test.py.

const line = (count: string) => SCORING_LINE_ITEMS.findIndex((l) => l.count === count);
const total = (state: GameState, seat: number) => scoreLines(state, seatOf(state, seat)!).reduce((a, b) => a + b, 0);

describe('the Score catalog', () => {
  it('carries the GS rows in TieBreakerPriority order, highest first', () => {
    expect(SCORING_LINE_ITEMS.map((l) => [l.id, l.multiplier, l.tieBreak])).toEqual([
      ['LINE_ITEM_ERA_BUILDINGS', 1, 1030],
      ['LINE_ITEM_ERA_SCORE', 1, 1010],
      ['LINE_ITEM_CIVICS', 3, 100],
      ['LINE_ITEM_CITIES', 5, 90],
      ['LINE_ITEM_DISTRICTS', 2, 80],
      ['LINE_ITEM_POPULATION', 1, 70],
      ['LINE_ITEM_GREAT_PEOPLE', 5, 60],
      ['LINE_ITEM_RELIGION', 5, 50],
      ['LINE_ITEM_TECHS', 2, 40],
      ['LINE_ITEM_WONDERS', 15, 30],
    ]);
    for (const l of SCORING_LINE_ITEMS) expect(l.categoryMultiplier).toBe(1);
  });
});

describe('the Score', () => {
  it('counts each line on a built state', () => {
    const state = makeState(makeMap(20, 12));
    const a = settleAt(state, 2 * 20 + 3);
    const b = settleAt(state, 8 * 20 + 14);
    const s = seatOf(state, 0)!;
    s.research.civics.push('CODE_OF_LAWS', 'CRAFTSMANSHIP');
    s.research.techs.push('POTTERY', 'MINING', 'ARCHERY');
    a.population = 3;
    b.population = 4;
    // a completed Campus counts, a placed Theater Square does not
    const campus = a.centerIndex + 1;
    a.districts.push({ type: 'CAMPUS', tileIndex: campus });
    state.map.tiles[campus].district = 'CAMPUS';
    state.map.tiles[campus].districtComplete = true;
    const theater = b.centerIndex + 1;
    b.districts.push({ type: 'THEATER_SQUARE', tileIndex: theater });
    state.map.tiles[theater].district = 'THEATER_SQUARE';
    // a completed wonder counts as a wonder AND as a district; one still
    // under construction counts as neither
    const pyr = a.centerIndex - 1;
    a.wonders.push({ id: 'PYRAMIDS', tileIndex: pyr });
    state.map.tiles[pyr].builtWonder = 'PYRAMIDS';
    state.map.tiles[pyr].builtWonderComplete = true;
    const ora = b.centerIndex - 1;
    b.wonders.push({ id: 'ORACLE', tileIndex: ora });
    state.map.tiles[ora].builtWonder = 'ORACLE';
    s.gpEarned.push(GREAT_PEOPLE.SCIENTIST[0].id, GREAT_PEOPLE.ENGINEER[0].id);
    // founded and enhanced: one belief of each of the four classes; the
    // pantheon is not one of them
    s.religion.pantheon = 'GOD_OF_THE_SEA';
    s.religion.founded = true;
    s.religion.follower = Object.keys(FOLLOWER_BELIEFS)[0];
    s.religion.founder = Object.keys(FOUNDER_BELIEFS)[0];
    s.religion.worship = Object.keys(WORSHIP_BELIEFS)[0];
    s.religion.enhancer = Object.keys(ENHANCER_BELIEFS)[0];
    s.eraScore = 9;
    s.eraScorePast = 30;
    // the capital's Palace, a Monument and a pillaged Granary are buildings;
    // the wonders are not
    a.buildings.push('MONUMENT');
    b.buildings.push('GRANARY');
    b.pillagedBuildings = ['GRANARY'];

    const got = scoreLines(state, s);
    expect(got[line('buildings')]).toBe(3);
    expect(got[line('eraScore')]).toBe(39);
    expect(got[line('civics')]).toBe(2 * 3);
    expect(got[line('cities')]).toBe(2 * 5);
    expect(got[line('districts')]).toBe(2 * 2);
    expect(got[line('population')]).toBe(7);
    expect(got[line('greatPeople')]).toBe(2 * 5);
    expect(got[line('religion')]).toBe(4 * 5);
    expect(got[line('techs')]).toBe(3 * 2);
    expect(got[line('wonders')]).toBe(15);
    expect(got.reduce((x, y) => x + y, 0)).toBe(3 + 39 + 6 + 10 + 4 + 7 + 10 + 20 + 6 + 15);
  });

  it('the era boundary banks the closed era into the whole game\'s era score', () => {
    const state = makeState(makeMap(20, 12));
    settleAt(state, 2 * 20 + 3);
    const s = seatOf(state, 0)!;
    s.eraScore = 12;
    s.eraScorePast = 5;
    state.turn = ERA_LENGTH;
    eraBoundary(state);
    expect(s.eraScore).toBe(0);
    expect(s.eraScorePast).toBe(17);
    expect(scoreLines(state, s)[line('eraScore')]).toBe(17);
  });
});

describe('the score victory', () => {
  function twoSeats() {
    const state = makeState(makeMap(24, 12));
    state.seats.push(emptySeat(1));
    settleAt(state, 3 * 24 + 4, 0);
    settleAt(state, 7 * 24 + 18, 1);
    return state;
  }

  it('names the seat with the highest Score', () => {
    const state = twoSeats();
    expect(scoreLeader(state)).toBe(0); // level: the lower seat
    seatOf(state, 1)!.research.techs.push('POTTERY');
    expect(scoreLeader(state)).toBe(1);
  });

  it('breaks a tie on the lines in TieBreakerPriority order, then to the lower seat', () => {
    const level = twoSeats();
    expect(total(level, 0)).toBe(total(level, 1));
    expect(scoreLeader(level)).toBe(0);
    // equal totals: seat 0 holds 2 more techs (+4), seat 1 4 more era score —
    // the era line (1010) outranks the techs (40)
    const era = twoSeats();
    seatOf(era, 0)!.research.techs.push('POTTERY', 'MINING');
    seatOf(era, 1)!.eraScore = (seatOf(era, 1)!.eraScore ?? 0) + 4;
    expect(total(era, 0)).toBe(total(era, 1));
    expect(scoreLeader(era)).toBe(1);
    // equal totals: seat 0 holds one more era score, seat 1 one more building
    // (+1) — the building line (1030) outranks the era (1010)
    const bld = twoSeats();
    seatOf(bld, 0)!.eraScore = (seatOf(bld, 0)!.eraScore ?? 0) + 1;
    seatOf(bld, 1)!.cities[0].buildings.push('MONUMENT');
    expect(total(bld, 0)).toBe(total(bld, 1));
    expect(scoreLeader(bld)).toBe(1);
    // equal totals: seat 0 holds 3 more techs (+6), seat 1 2 more civics (+6)
    // — the civic line (100) outranks the techs (40)
    const civ = twoSeats();
    seatOf(civ, 0)!.research.techs.push('POTTERY', 'MINING', 'ARCHERY');
    seatOf(civ, 1)!.research.civics.push('CODE_OF_LAWS', 'CRAFTSMANSHIP');
    expect(total(civ, 0)).toBe(total(civ, 1));
    expect(scoreLeader(civ)).toBe(1);
  });

  it('a seat with no city cannot win it', () => {
    const state = twoSeats();
    seatOf(state, 1)!.research.techs.push('POTTERY', 'MINING', 'ARCHERY');
    seatOf(state, 1)!.cities = [];
    expect(scoreLeader(state)).toBe(0);
    seatOf(state, 0)!.cities = [];
    expect(scoreLeader(state)).toBe(-1);
  });

  it('past TURN_LIMIT endTurn ends the game on the score and names the leader', () => {
    const state = seededGame(4242, 2);
    state.autoResearch = false;
    seatOf(state, 1)!.research.civics.push('CODE_OF_LAWS', 'CRAFTSMANSHIP', 'FOREIGN_TRADE', 'EARLY_EMPIRE');
    state.turn = TURN_LIMIT - 1;
    endTurn(state);
    expect(state.gameOver).toBe(false);
    expect(state.victoryRow).toBe(-1);
    endTurn(state);
    expect(state.gameOver).toBe(true);
    expect(state.victoryType).toBe(1);
    expect(state.victoryRow).toBe(1);
    expect(state.victoryRow).toBe(scoreLeader(state));
  });
});
